### PR TITLE
[BACKLOG-4381] - REST interfaces for User and Roles Create, Retrieve,…

### PR DESCRIPTION
**Pull request by brendan-lynch**
… Update, and Delete Functionality

Refactored REST endpoints and created unit test coverage.

### CODE REVIEW

File: extensions/src/org/pentaho/platform/web/http/api/resources/services/UserRoleDaoService.java [hunk 1]
```diff
@@ -17,15 +17,149 @@
...
+import java.util.StringTokenizer;
+
 public class UserRoleDaoService {
-    public UserListWrapper getUsers() throws Exception {
-        IUserRoleDao roleDao =
-                PentahoSystem.get( IUserRoleDao.class, "userRoleDaoProxy", PentahoSessionHolder.getSession() );
-        return new UserListWrapper( roleDao.getUsers() );
+  public UserListWrapper getUsers() throws Exception {
+    IUserRoleDao roleDao =
+      PentahoSystem.get( IUserRoleDao.class, "userRoleDaoProxy", PentahoSessionHolder.getSession() );
```
Comment (pentaho-nbaker) [inline extensions/src/org/pentaho/platform/web/http/api/resources/services/UserRoleDaoService.java]: Replace this with simply PentahoSystem.get( IUserRoleDao.class );

Also cache the userRoleDao as a member variable of this class so you're not looking it up every time.

File: extensions/src/org/pentaho/platform/web/http/api/resources/services/UserRoleDaoService.java [hunk 2]
```diff
@@ -17,15 +17,149 @@
...
+    if ( canAdminister() ) {
+      IUserRoleDao roleDao =
+        PentahoSystem.get( IUserRoleDao.class, "userRoleDaoProxy", PentahoSessionHolder.getSession() );
+      return new UserListWrapper( roleDao.getRoleMembers( getTenant(), roleName ) );
+    } else {
+      throw new SecurityException();
+    }
+  }
+
+  public void deleteUser( String userNames )
```
Comment (pentaho-nbaker) [inline extensions/src/org/pentaho/platform/web/http/api/resources/services/UserRoleDaoService.java]: Pluralize this.

File: extensions/src/org/pentaho/platform/web/http/api/resources/services/UserRoleDaoService.java [hunk 3]
```diff
@@ -17,15 +17,149 @@
...
+        if ( user != null ) {
+          roleDao.deleteUser( user );
+        }
+      }
+    } else {
+      throw new SecurityException();
+    }
+  }
+
+  public void deleteRole( String roleNames ) throws SecurityException, UncategorizedUserRoleDaoException {
```
Comment (pentaho-nbaker) [inline extensions/src/org/pentaho/platform/web/http/api/resources/services/UserRoleDaoService.java]: Pluralize

File: extensions/src/org/pentaho/platform/web/http/api/resources/services/UserRoleDaoService.java [hunk 4]
```diff
@@ -17,15 +17,149 @@
...
+          roleDao.deleteRole( role );
+        }
+      }
+    } else {
+      throw new SecurityException();
+    }
+  }
+
+  private boolean canAdminister() {
+    IAuthorizationPolicy policy = PentahoSystem.get( IAuthorizationPolicy.class );
```
Comment (pentaho-nbaker) [inline extensions/src/org/pentaho/platform/web/http/api/resources/services/UserRoleDaoService.java]: Cache the IAuthorizationPolicy

File: extensions/src/org/pentaho/platform/web/http/api/resources/services/UserRoleDaoService.java [hunk 5]
```diff
@@ -17,15 +17,149 @@
...
+    return policy.isAllowed( RepositoryReadAction.NAME ) && policy.isAllowed( RepositoryCreateAction.NAME )
+      && ( policy.isAllowed( AdministerSecurityAction.NAME ) );
+  }
+
+  private ITenant getTenant() {
```
Comment (pentaho-nbaker) [inline extensions/src/org/pentaho/platform/web/http/api/resources/services/UserRoleDaoService.java]: Use org.pentaho.platform.engine.core.system.TenantUtils#getCurrentTenant instead

**Comment by pentaho-nbaker** [comment]
I'm pretty terribly disappointed that we're using tab delimited strings in-place of arrays throughout this code. I understand that's the way it's been and we said we didn't want to change the payloads... Lets leave it for now and if we get time we'll revisit it before release.

**Comment by brendan-lynch** [comment]
@CodeOnCoffee Thanks for the review, I updated the code from the review, including the resource layer class.

**Comment by rfellows** [comment]
Looks good

**Pull request merged**